test/test_sparse_csr.py

```diff
@@ -585,6 +585,71 @@ def test_csr_matvec(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, err_msg):
                 csr.matmul(bad_vec)
 
+    @onlyCUDA
+    @skipCUDAIfNoCusparseGeneric
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_baddbmm(self, device, dtype):
+        def run_test(c, a, a_batched, b, op_b=False, op_out=False, *, dtype=None, device=None):
+            alpha = complex(random.random(), random.random()) if dtype.is_complex else random.random()
+            beta = complex(random.random(), random.random()) if dtype.is_complex else random.random()
+            b = b.mH if (op_b and a.shape == b.shape) else b
+
+            actual = torch.baddbmm(c, a_batched, b, alpha=alpha, beta=beta)
+
+            out = torch.empty_like(c.mH if op_out and a.shape == b.shape else c)
+            torch.baddbmm(c, a_batched, b, alpha=alpha, beta=beta, out=out)
+
+            expected = [torch.addmm(c[i], a, b[i], alpha=alpha, beta=beta) for i in range(c.shape[0])]
+            expected = torch.stack(expected, 0)
+
+            self.assertEqual(actual, out)
+            self.assertEqual(actual, expected)
+
+        for index_dtype in [torch.int32, torch.int64]:
+            for (m, n, k), batch_size, noncontiguous in zip(itertools.product([1, 5], repeat=3), [1, 3], [True, False]):
+                nnz = random.randint(0, m * k)
+                a = self.genSparseCSRTensor((m, k), nnz, dtype=dtype, device=device, index_dtype=index_dtype)
+
+                # a_batched is a regular CSR tensor but with a batch dimension in the shape
+                a_batched = torch._sparse_csr_tensor_unsafe(
```

Comment:
So this is effectively a "hack" to circumvent the shape check that baddmm requires, because in the underlying C++ code you constraint yourself to the first batch entry of the CSR matrix? I guess if we allowed broadcasting in the batch dimensions this wouldn't be necessary and batch_size could be 1.

Reply:
Right, it's a hack to circumvent the shape check. `matmul` allows broadcasting batch dimension, but `bmm` doesn't.
```
In [1]: import torch
In [2]: a = torch.randn(3, 3)
In [3]: b = torch.randn(2, 3, 3)
In [4]: a @ b
Out[4]:
tensor([[[-2.3947,  4.1253,  1.6287],
         [-0.8864,  0.6599, -0.7547],
         [ 5.3741, -4.8693,  2.2964]],

        [[ 4.5006,  0.3885, -2.6066],
         [ 0.4729,  0.0400, -0.4279],
         [-5.1257, -0.6420,  3.7720]]])
In [5]: a.bmm(b)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-bd5c7f2de399> in <module>
----> 1 a.bmm(b)

RuntimeError: batch1 must be a 3D tensor
In [6]: a.unsqueeze(0).bmm(b)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-5dac428fe979> in <module>
----> 1 a.unsqueeze(0).bmm(b)

RuntimeError: Expected size for first two dimensions of batch2 tensor to be: [1, 3] but got: [2, 3].
```
I need to do something so that this hack is not required.